### DISCUSSION TITLE
A 'DarcyPressure' is not a regidtered object.

### DISCUSSION THREAD

**Discussion by Yaode-Yin** [idaholab/moose#25626] [Q&A General]
I am a newcomer to MOOSE and have written the Babber App on  the tutorials  'Step 5: Develop a Kernel Object' : https://mooseframework.inl.gov/getting_started/examples_and_tutorials/tutorial01_app_development/step05_kernel_object.html#darcy-source
I do the same things as the tutorials, but when I execute the input file, it cannot works.
```
*** ERROR ***
A 'DarcyPressure' is not a registered object.
........

I dont know why the 'DarcyPressure' cannot  be registered correctly. 
DarcyPressure.h in include/kernel

#pragma once

// Including the "ADKernel" base class here so we can extend it
#include "ADKernelGrad.h"

/**
 * Computes the residual contribution: K / mu * grad_test * grad_u.
 */
class DarcyPressure : public ADKernelGrad
{
public:
  static InputParameters validParams();

  DarcyPressure(const InputParameters & parameters);

protected:
  /// ADKernel objects must override precomputeQpResidual
  virtual ADRealVectorValue precomputeQpResidual() override;

  /// The variables which hold the value for K and mu
  const Real _permeability;
  const Real _viscosity;
};

DarcyPressure.C in src/kernel

#include "DarcyPressure.h"

registerMooseObject("BabblerApp", DarcyPressure);

InputParameters
DarcyPressure::validParams()
{
	InputParameters params = ADKernelGrad::validParams();
	params.addClassDescription("Compute the diffusion term for Darcy pressure ($p$) equation: "
		"$-\\nabla \\cdot \\frac{\\mathbf{K}}{\\mu} \\nabla p = 0$");

	return params;
}

DarcyPressure::DarcyPressure(const InputParameters & parameters)
	: ADKernelGrad(parameters),

	// Set the coefficients for the pressure kernel
	_permeability(0.8451e-09),
	_viscosity(7.98e-04)
{
}

ADRealVectorValue
DarcyPressure::precomputeQpResidual()
{
	return (_permeability / _viscosity) * _grad_u[_qp];
}

 ```

**Comment by GiudGiud** [comment]
Hello

What is the name of the app in the src/base folder?
There's an XXXApp named file. Then if you open that file you can see more.

We're looking for issues with capitalization

Guillaume

**Reply by Yaode-Yin** [reply]
Hello 

Its name is BabblerApp.C . Here's the codes

#include "DarcyPressure.h"

registerMooseObject("BabblerApp", DarcyPressure);

InputParameters
DarcyPressure::validParams()
{
	InputParameters params = ADKernelGrad::validParams();
	params.addClassDescription("Compute the diffusion term for Darcy pressure ($p$) equation: "
		"$-\\nabla \\cdot \\frac{\\mathbf{K}}{\\mu} \\nabla p = 0$");

	return params;
}

DarcyPressure::DarcyPressure(const InputParameters & parameters)
	: ADKernelGrad(parameters),

	// Set the coefficients for the pressure kernel
	_permeability(0.8451e-09),
	_viscosity(7.98e-04)
{
}

ADRealVectorValue
DarcyPressure::precomputeQpResidual()
{
	return (_permeability / _viscosity) * _grad_u[_qp];
}

**Reply by GiudGiud** [reply]
can you copy-paste babblerApp.C here?

also you compiled after adding this object right?

**Reply by Yaode-Yin** [reply]
Sorry, not thar code

**Reply by Yaode-Yin** [reply]
#include "BabblerApp.h"
#include "Moose.h"
#include "AppFactory.h"
#include "ModulesApp.h"
#include "MooseSyntax.h"

InputParameters
BabblerApp::validParams()
{
  InputParameters params = MooseApp::validParams();
  params.set<bool>("use_legacy_material_output") = false;
  return params;
}

BabblerApp::BabblerApp(InputParameters parameters) : MooseApp(parameters)
{
  BabblerApp::registerAll(_factory, _action_factory, _syntax);
}

BabblerApp::~BabblerApp() {}

void 
BabblerApp::registerAll(Factory & f, ActionFactory & af, Syntax & s)
{
  ModulesApp::registerAllObjects<BabblerApp>(f, af, s);
  Registry::registerObjectsTo(f, {"BabblerApp"});
  Registry::registerActionsTo(af, {"BabblerApp"});

  /* register custom execute flags, action syntax, etc. here */
}

void
BabblerApp::registerApps()
{
  registerApp(BabblerApp);
}

**Reply by Yaode-Yin** [reply]
The code seems to be  automatically generated by Moose.

**Reply by GiudGiud** [reply]
seems fine

ok let s change this line 
registerMooseObject("BabblerApp", DarcyPressure);
to this
registerMooseObject("MooseApp", DarcyPressure);
and recompile please

**Reply by Yaode-Yin** [reply]
After 
```
cd ~/projects/babbler 
make -j4
cd ~/projects/babbler/problems
../babbler-opt -i pressure_diffusion.i 
```

it still not work correctly.
'DarcyPressure' is not a registered object. 
Do I have Installation failure?

**Reply by Yaode-Yin** [reply]
Of course, I've changed the code

**Reply by GiudGiud** [reply]
What does `make` return? Does it give an error?

**Reply by Yaode-Yin** [reply]
No, make has no error.

**Reply by GiudGiud** [reply]
in babbler, what does:
`ls -l *`
and `ls -l */*`
return?

**Reply by Yaode-Yin** [reply]
```
~/projects/babbler$ ls -1 *
babbler-opt
babbler.yaml
LICENSE
Makefile
README.md
run_tests
testroot

build:
header_symlinks
unity_src

doc:
config.yml
content
moosedocs.py

include:
base
kernels

lib:
libbabbler-opt.la
libbabbler-opt.so
libbabbler-opt.so.0
libbabbler-opt.so.0.0.0

problems:
pressure_diffusion.i

scripts:
install-format-hook.sh

src:
base
kernels
main.C
main.x86_64-conda-linux-gnu.opt.lo
main.x86_64-conda-linux-gnu.opt.lo.d

test:
include
lib
src
tests

unit:
include
Makefile
run_tests
src




~/projects/babbler$ ls -1 */*
doc/config.yml
doc/moosedocs.py
lib/libbabbler-opt.la
lib/libbabbler-opt.so
lib/libbabbler-opt.so.0
lib/libbabbler-opt.so.0.0.0
problems/pressure_diffusion.i
scripts/install-format-hook.sh
src/main.C
src/main.x86_64-conda-linux-gnu.opt.lo
src/main.x86_64-conda-linux-gnu.opt.lo.d
unit/Makefile
unit/run_tests

build/header_symlinks:
BabblerApp.h
BabblerTestApp.h
DarcyPressure.h

build/unity_src:
kernels_Unity.C
kernels_Unity.x86_64-conda-linux-gnu.opt.lo
kernels_Unity.x86_64-conda-linux-gnu.opt.lo.d

doc/content:
index.md

include/base:
BabblerApp.h

include/kernels:
DarcyPressure.h

src/base:
BabblerApp.C
BabblerApp.x86_64-conda-linux-gnu.opt.lo
BabblerApp.x86_64-conda-linux-gnu.opt.lo.d

src/kernels:
DarcyPressure.C

test/include:
base

test/lib:
libbabbler_test-opt.la
libbabbler_test-opt.so
libbabbler_test-opt.so.0
libbabbler_test-opt.so.0.0.0

test/src:
base

test/tests:
kernels

unit/include:
place_holder

unit/src:
main.C
SampleTest.C
```

**Reply by GiudGiud** [reply]
ok no issues there.

try this to save your work and get a fresh build
```
git commit -am "First commit"
cd ~/projects/babbler
git clean -xfd
make -j 4
```

**Reply by Yaode-Yin** [reply]
It still not work. Mayne something wrong with my  installation? I have tried to delete babbler folder, and recreate it with 'stork' command. But DarcyPressure still cannot be registered.

**Reply by GiudGiud** [reply]
are you on linux or mac?
what does this return:
```
ldd babbler-opt                       /.         otool -L babbler-opt
```
on linux / mac
